### PR TITLE
AllTestsAdded

### PR DESCRIPTION
**Pull request by SamDotNetDev**

### CODE REVIEW

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+                Address = "Test Address",
+            };
+        }
+
+        #region Index
+        [Fact]
+        public void GivenNothing_WhenIndexIsCalled_ThenReturnsViewResult()
+        {
+            //Arrange
+            var expectedParkingZones = new List<ParkingZone>() { _parkingZoneTest };
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]:
```suggestion
            var expectedParkingZones = new List<ViewModel>() { _parkingZoneTest };
```

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+            _service.Setup(x => x.GetById(Id)).Returns(_parkingZoneTest);
+
+            //Act
+            var viewresult = _controller.Details(Id);
+
+            //Assert
+            var model = Assert.IsType<ViewResult>(viewresult).Model;
+            Assert.NotNull(viewresult);
+            Assert.IsType<DetailsVM>(model);
+            Assert.IsAssignableFrom<DetailsVM>(model);
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]: remove one of them

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+            //Act
+            var viewresult = _controller.Create();
+
+            //Assert
+            Assert.IsType<ViewResult>(viewresult);
+        }
+
+        [Fact]
+        public void GivenViewModel_WhenCreateIsCalledToPost_ThenModelStateIsFalseReturnsVM()
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]:
```suggestion
        public void GivenCreateViewModel_WhenCreateIsCalledToPost_ThenModelStateIsFalseAndReturnsVM()
```

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        {
+            //Arrange
+            CreateVM createVM = new();
+            _controller.ModelState.AddModelError("Name", "Required");
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]: Inline data

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+            //Assert\
+            var model = Assert.IsType<ViewResult>(result).Model;
+            Assert.IsType<CreateVM>(model);
+            Assert.NotNull(createVM);
+            Assert.False(_controller.ModelState.IsValid);
+        }
+
+        [Fact]
+        public void GivenViewModel_WhenCreateIsCalled_ThenModelStateIsTrueReturnsRedirectToAction()
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]:
```suggestion
        public void GivenCreateVM_WhenCreateIsCalled_ThenModelStateIsTrueAndReturnsRedirectToAction()
```

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        public void GivenParkingZoneId_WhenEditIsCalled_ThenReturnsVM()
+        {
+            //Arrange
+            var ExpectedVM = new EditVM()
+            {
+                Id = Id,
+                Name = "Test",
+                Address = "Test Address"
+            };
+            _service.Setup(x=>x.GetById(Id)).Returns(_parkingZoneTest);
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]:
```suggestion
            _service.Setup(x => x.GetById(Id)).Returns(_parkingZoneTest);
```

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            
+            //Act
+            var result = _controller.Edit(Id, editVM);
+
+            //Assert
+            Assert.IsType<NotFoundResult>(result);
+        }
+
+        [Fact]
+        public void GivenIdAndEditVM_WhenEditIsCalledToPost_ThenModelStateIsTrueReturnsRedirectToAction()
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]:
```suggestion
        public void GivenIdAndEditVM_WhenEditIsCalledToPost_ThenModelStateIsTrueAndReturnsRedirectToAction()
```

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        {
+            //Arrange
+            EditVM editVM = new() { Id = Id};
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]:
```suggestion
            EditVM editVM = new() { Id = Id };
```

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+            //Act
+            var result = _controller.Edit(Id, editVM);
+
+            //Assert
+            Assert.IsType<RedirectToActionResult>(result);
+            Assert.True(_controller.ModelState.IsValid);
+        }
+
+        [Fact]
+        public void GivenIdAndEditVM_WhenEditIsCalledToPost_ThenModelStateIsFalseReturnsView()
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]:
```suggestion
        public void GivenIdAndEditVM_WhenEditIsCalledToPost_ThenModelStateIsFalseAndReturnsView()
```

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+        {
+            //Arrange
+            EditVM editVM = new() { Id=Id};
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]:
```suggestion
            EditVM editVM = new() { Id=Id };
```

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+            _controller.ModelState.AddModelError("Name", "Required");
+            _controller.ModelState.AddModelError("Address", "Required");
+            _controller.ModelState.AddModelError("DateOfEstablishment", "Required");
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]: [Inline Data], [Member data]

File: ParkingZoneTest/Services/ParkingZoneTests.cs [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+using ParkingZoneApp;
+using ParkingZoneApp.Services;
+using ParkingZoneApp.Repositories;
+using ParkingZoneApp.Models;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.VisualStudio.Web.CodeGenerators.Mvc.Templates.BlazorIdentity.Pages;
+
+namespace ParkingZoneTest.Services
+{
+    public class ParkingZoneTests
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Services/ParkingZoneTests.cs]:
```suggestion
    public class ParkingZoneServiceTests
```

File: ParkingZoneTest/Controllers/ParkingZoneTests.cs [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+using System.Threading.Tasks;
+using System.Xml.Serialization;
+
+namespace ParkingZoneTest.Controllers
+{
+    public class ParkingZoneTests
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneTests.cs]:
```suggestion
    public class ParkingZoneControllerTests
```

File: ParkingZoneTest/Services/ParkingZoneTests.cs [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+            _repository.Verify(x => x.Delete(_ParkingZoneTest), Times.Once());
+        }
+        #endregion
+
+        #region GetAll
+        [Fact]
+        public void GivenNothing_WhenGetAllIsCalled_ThenRepositoryGetAllIsCalled()
+        {
+            var parkingZones = new List<ParkingZone>();
+            // Arrange
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Services/ParkingZoneTests.cs]:
```suggestion
            // Arrange
            var parkingZones = new List<ParkingZone>();
```

File: ParkingZoneTest/Services/ParkingZoneTests.cs [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+            _repository.Setup(repo => repo.GetById(Id)).Returns(_ParkingZoneTest);
+
+            // Act
+            var result = _service.GetById(Id);
+
+            // Assert
+            Assert.NotNull(result);
+            Assert.IsType<ParkingZone>(result);
+            _repository.Verify(x => x.GetById(Id), Times.Once());
+        }
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Services/ParkingZoneTests.cs]: JsonSerializer

**Review by abdullokhtolibjonov — CHANGES_REQUESTED** [review]

File: ParkingZoneTest/Controllers/ParkingZoneControllerTests.cs [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+                Address = "Test Address",
+            };
+        }
+
+        #region Index
+        [Fact]
+        public void GivenNothing_WhenIndexIsCalled_ThenReturnsViewResult()
+        {
+            //Arrange
+            var expectedVMs = new List<ListItemVM>() {new (_parkingZoneTest) };
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneControllerTests.cs]:
```suggestion
            var expectedVMs = new List<ListItemVM>() { new (_parkingZoneTest) };
```

**Review by abdullokhtolibjonov — CHANGES_REQUESTED** [review]

File: ParkingZoneTest/Controllers/ParkingZoneControllerTests.cs [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+            //Act
+            var viewresult = _controller.Create();
+
+            //Assert
+            Assert.IsType<ViewResult>(viewresult);
+        }
+
+        [Theory]
+        [InlineData("Name","Required")]
+        [InlineData("Address","Required")]
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneControllerTests.cs]:
```suggestion
```

File: ParkingZoneTest/Controllers/ParkingZoneControllerTests.cs [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+
+            //Assert
+            Assert.IsType<RedirectToActionResult>(result);
+            Assert.True(_controller.ModelState.IsValid);
+        }
+
+        [Theory]
+        [InlineData("Name","Required")]
+        [InlineData("Address","Required")]
+        [InlineData("DateOfEstablishment", "Required")]
```
Comment (abdullokhtolibjonov) [inline ParkingZoneTest/Controllers/ParkingZoneControllerTests.cs]:
```suggestion
```

**Review by abdullokhtolibjonov — CHANGES_REQUESTED** [review]

**Review by abdullokhtolibjonov — APPROVED** [review]

File: ParkingZoneApp/Areas/Admin/Views/ParkingZone/Edit.cshtml [hunk 19]
```diff
@@ -15,6 +15,7 @@
         <form asp-action="Edit">
             <div asp-validation-summary="ModelOnly" class="text-danger"></div>
             <input type="hidden" asp-for="Id" />
+            <input type="hidden" asp-for="DateOfEstablishment"/>
```
Comment (napster1991) [inline ParkingZoneApp/Areas/Admin/Views/ParkingZone/Edit.cshtml:18]: No need in this property on Edit view. Remove it.

File: ParkingZoneApp/Models/ParkingZone.cs [hunk 20]
```diff
@@ -8,7 +8,7 @@ public class ParkingZone
     {
         [Key]
         [Required]
-        public int Id { get; set; }
+        public int? Id { get; set; }
```
Comment (napster1991) [inline ParkingZoneApp/Models/ParkingZone.cs:11]: Revert! Id must not be nullable!

File: ParkingZoneApp/Models/ParkingZone.cs [hunk 21]
```diff
@@ -17,7 +17,7 @@ public class ParkingZone
         public string Address { get; set; }
 
         [Required]
-        public DateTime DateOfEstablishment { get; set; }
+        public DateTime? DateOfEstablishment { get; set; }
```
Comment (napster1991) [inline ParkingZoneApp/Models/ParkingZone.cs:20]: The same.

File: ParkingZoneApp/ViewModels/ParkingZones/DetailsVM.cs [hunk 22]
```diff
@@ -1,22 +1,27 @@
...
 {
     public class DetailsVM
     {
-        public DetailsVM()
-        {
-            
-        }
+
+        [Required]
+        public int? Id { get; set; }
```
Comment (napster1991) [inline ParkingZoneApp/ViewModels/ParkingZones/DetailsVM.cs:10]: Id must not be nullable!

File: ParkingZoneApp/ViewModels/ParkingZones/EditVM.cs [hunk 23]
```diff
@@ -1,24 +1,32 @@
...
 {
     public class EditVM
     {
-        public EditVM()
-        {
-            
-        }
+
+        [Required]
+        public int? Id { get; set; }
```
Comment (napster1991) [inline ParkingZoneApp/ViewModels/ParkingZones/EditVM.cs:10]: The same.

File: ParkingZoneApp/ViewModels/ParkingZones/EditVM.cs [hunk 24]
```diff
@@ -1,24 +1,32 @@
...
+
+        [Required]
+        public string Name { get; set; }
+
+        [Required]
+        public string Address { get; set; }
+
+        [Required]
+        public DateTime? DateOfEstablishment { get; set; }
```
Comment (napster1991) [inline ParkingZoneApp/ViewModels/ParkingZones/EditVM.cs:19]: This property is uneditable. Remove it.

File: ParkingZoneApp/ViewModels/ParkingZones/ListItemVM.cs [hunk 25]
```diff
@@ -1,19 +1,30 @@
 ﻿using ParkingZoneApp.Models;
+using System.ComponentModel.DataAnnotations;
 
 namespace ParkingZoneApp.ViewModels.ParkingZones
 {
     public class ListItemVM
     {
+        [Required]
+        public int? Id { get; set; }
```
Comment (napster1991) [inline ParkingZoneApp/ViewModels/ParkingZones/ListItemVM.cs:9]: Not nullable.

File: ParkingZoneApp/ViewModels/ParkingZones/ListItemVM.cs [hunk 26]
```diff
@@ -1,19 +1,30 @@
...
+
+        [Required]
+        public string Name { get; set; }
+
+        [Required]
+        public string Address { get; set; }
+
+        [Required]
+        public DateTime? DateOfEstablishment { get; set; }
```
Comment (napster1991) [inline ParkingZoneApp/ViewModels/ParkingZones/ListItemVM.cs:18]: Remove this property.

File: ParkingZoneTest/ModelTests/CreateVMValidationTests.cs [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace ParkingZoneTest.ModelTests
+{
+    public class CreateVMValidationTests
+    {
+        public static IEnumerable<object[]> TestData =>
+        new List<object[]>
+        {
+            new object[] { null, "Test Address" },
+            new object[] { "Test2", null }
```
Comment (napster1991) [inline ParkingZoneTest/ModelTests/CreateVMValidationTests.cs:12]: I would also add a valid case and a third property (for all cases) indicating if the model is valid: `new object[] { "Test2", "test address", true }`

File: ParkingZoneTest/ModelTests/CreateVMValidationTests.cs [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+        };
+
+        [Theory]
+        [MemberData(nameof(TestData))]
+        public void GivenInvalidItem_WhenCreatingCreateVM_ThenValidationIsFalse(string Name, string Address)
```
Comment (napster1991) [inline ParkingZoneTest/ModelTests/CreateVMValidationTests.cs:17]: My suggestion:

`public void GivenItemToBeValidated_WhenCreatingCreateVM_ThenValidationIsPerformed(string Name, string Address, bool expectedValidationResult)`

File: ParkingZoneTest/ModelTests/CreateVMValidationTests.cs [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+                Address = Address
+            };
+            var validationContext = new ValidationContext(createVM, null, null);
+            var validationResult = new List<ValidationResult>();
+
+            //Act
+            var result = Validator.TryValidateObject(createVM,validationContext,validationResult);
+
+            //Assert
+            Assert.False(result);
```
Comment (napster1991) [inline ParkingZoneTest/ModelTests/CreateVMValidationTests.cs:32]: `Assert.IsEqual(expectedValidationResult, result);`

File: ParkingZoneTest/ModelTests/CreateVMValidationTests.cs [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+            Assert.NotEmpty(validationResult);
```
Comment (napster1991) [inline ParkingZoneTest/ModelTests/CreateVMValidationTests.cs:33]: This assertion can be omitted.

**Pull request merged**